zk/datadog_checks/zk/zk.py

```diff
@@ -220,33 +223,49 @@ def report_instance_mode(self, mode):
             gauge_name = 'zookeeper.instances.%s' % k
             self.gauge(gauge_name, v)
 
-    def _send_command(self, command):
-        buf = StringIO()
+    @staticmethod
+    def _get_data(sock, command):
         chunk_size = 1024
         max_reads = 10000
-        # try-finally and try-except to stay compatible with python 2.4
-        with closing(socket.socket()) as sock:
-            sock.settimeout(self.timeout)
-            try:
-                # Connect to the zk client port and send the stat command
-                sock.connect((self.host, self.port))
-                sock.sendall(ensure_bytes(command))
-
-                # Read the response into a StringIO buffer
-                chunk = ensure_unicode(sock.recv(chunk_size))
-                buf.write(chunk)
-                num_reads = 1
-
-                while chunk:
-                    if num_reads > max_reads:
-                        # Safeguard against an infinite loop
-                        raise Exception("Read %s bytes before exceeding max reads of %s. " % (buf.tell(), max_reads))
-                    chunk = ensure_unicode(sock.recv(chunk_size))
-                    buf.write(chunk)
-                    num_reads += 1
-            except (socket.timeout, socket.error):
-                raise ZKConnectionFailure()
+        buf = StringIO()
+        sock.sendall(ensure_bytes(command))
+        # Read the response into a StringIO buffer
+        chunk = ensure_unicode(sock.recv(chunk_size))
+        buf.write(chunk)
+        num_reads = 1
+
+        while chunk:
+            if num_reads > max_reads:
+                # Safeguard against an infinite loop
+                raise Exception("Read %s bytes before exceeding max reads of %s. " % (buf.tell(), max_reads))
+            chunk = ensure_unicode(sock.recv(chunk_size))
+            buf.write(chunk)
+            num_reads += 1
+        return buf
 
+    def _send_command(self, command):
+        buf = StringIO()
+        if self.enable_ssl:
```

Comment:
The two branches of that if can be more similar. You should be able to do something like this
```python
# AF_INET and SOCK_STREAM are the default but being explicit here helps for such a low-level lib
with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
    # Timeout can be set on the initial sock, it'll be passed to ssock.
    sock.settimeout(self.timeout)
    try:
        if self.enable_ssl:
            # You need a `with` as well for the sslsocket, otherwise it's not closed.
            with closing(context.wrap_socket(sock, server_hostname=self.host)) as ssock:
                return self._get_data(ssock, command)
        else:
            return self._get_data(sock, command)
    except (socket.timeout, socket.error):
        raise ZKConnectionFailure()
    
```

Reply:
🔥